crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1626,11 +1674,62 @@ impl Client {
         &self,
         alias: &RoomOrAliasId,
         server_names: &[OwnedServerName],
-    ) -> HttpResult<join_room_by_id_or_alias::v3::Response> {
+    ) -> Result<room::Joined> {
         let request = assign!(join_room_by_id_or_alias::v3::Request::new(alias), {
             server_name: server_names,
         });
-        self.send(request, None).await
+
+        let (tx, mut rx) = mpsc::channel::<Result<room::Joined>>(1);
+        let (tx_room_id, rx_room_id) = watch::channel::<Option<OwnedRoomId>>(None);
+
+        self.add_event_handler({
+            move |event: SyncStateEvent<RoomMemberEventContent>,
+                  room: room::Room,
+                  client: Client,
+                  handle: EventHandlerHandle| {
+                let tx = tx.clone();
+                let mut rx_room_id = rx_room_id.clone();
+
+                async move {
+                    while rx_room_id.borrow_and_update().is_none() {
+                        rx_room_id.changed().await.expect("receive error for room_id");
+                    }
+
+                    let room_id =
+                        rx_room_id.borrow().to_owned().expect("room_id not provided by channel");
+
+                    if event.membership() == &MembershipState::Join
+                        && event.state_key().as_str() == client.user_id().expect("user_id").as_str()
+                        && room.room_id() == room_id
+                    {
+                        debug!("received RoomMemberEvent corresponding to requested join");
+
+                        client.remove_event_handler(handle).await;
+
+                        let joined_result = if let room::Room::Joined(joined_room) = room {
+                            Ok(joined_room)
+                        } else {
+                            warn!("Corresponding Room not in state: joined");
+                            Err(Error::InconsistentState)
+                        };
+
+                        if let Err(e) = tx.send(joined_result).await {
+                            debug!(
+                                "Sending event from event_handler failed, \
+                                 receiver already dropped: {}",
+                                e
+                            );
+                        }
+                    }
+                }
+            }
+        })
+        .await;
+
+        let response = self.send(request, None).await?;
```

Comment:
In case this fails, we should also unregister the event handler.

Reply:
We are doing that in line 1708, are we not?

Reply:
Yes, kinda. In any case, a drop guard is probably better. Currently working on that.